src/goose/synopsis/system.py

```diff
@@ -0,0 +1,149 @@
+import json
+from tiktoken import get_encoding
+from exchange import Message
+import subprocess
+import os
+import atexit
+import platform
+from pathlib import Path
+from typing import Dict, Iterable, List, Set
+
+from attrs import define, field
+from exchange.content import ToolUse
+from goose.toolkit.utils import get_language
+
+
+@define
+class File:
+    path: str
+    content: str
+    language: str
+
+    @property
+    def context(self) -> str:
+        return f"""\
+{self.path}
+```{self.language}
+{self.content}
+```"""
+
+
+@define
+class OperatingSystem:
+    """
+    OperatingSystem class that can manage background processes created using subprocess.Popen.
+    """
+
+    cwd: str = os.getcwd()
+    platform: str = platform.system()
+    env: Dict[str, str] = os.environ.copy()
+    _active_files: Set[str] = field(init=False, factory=set)
+    _processes: Dict[int, subprocess.Popen] = field(init=False, factory=dict)
+
+    def __attrs_post_init__(self) -> None:
+        atexit.register(self._cleanup_processes)
+
+    def _cleanup_processes(self) -> None:
+        """Terminate all processes in the process queue."""
+        for process_id in list(self._processes.keys()):
+            self.cancel_process(process_id)
+
```

Comment:
we can save this for later, but i feel adding async context management might be useful for better handling of process lifecycles.

```python3
@asynccontextmanager
async def managed_process(*args, timeout=None, **kwargs):
    process = await asyncio.create_subprocess_exec(*args, **kwargs)
    try:
        try:
            await asyncio.wait_for(
                asyncio.create_task(process.wait()),
                timeout=timeout,
            )
        except Exception:
            print(
                f"timeout limit {timeout} second(s) exceeded, "
                f"terminating process (pid={process.pid}, args={args})"
            )
            process.terminate()
            await process.wait()
        yield process
    finally:
        if process.returncode is None:
            print(f"cleaning up process: {process.pid}")
            process.terminate()
            await process.wait()
        else if process.returncode != 0:
            # clean up
...

async def main():
    async with managed_process("sleep", "600", timeout=5) as process:
        pass
```

the error also becomes consistent for us to handle as it's normalized across all commands
```sh
timeout limit 5 second(s) exceeded, terminating process (pid=93675, args=('sleep', '600'))
```

it's a bit extra but now we can 
 1. forcibly exit long running commands (force another eval loop)
 2. additionally, this would allow creating some "chunking" workflows 
   * i often will pipe a series of commands to whittle down a surface area, creating intermediate files to make limit the scope of downstream toolchains